### PR TITLE
Make signature of Path::strip_prefix accept non-references

### LINKED ISSUE #48390: The signature of Path::strip_prefix is bizarrely overconstrained
```rust
pub fn strip_prefix<'a, P: ?Sized>(
    &'a self, 
    base: &'a P  // <===  ???  a reference?
) -> Result<&'a Path, StripPrefixError> 
where P: AsRef<Path>, 
```

This can clearly be generalized to

```rust
pub fn strip_prefix<'a, P>(
    &'a self, 
    base: P
) -> Result<&'a Path, StripPrefixError> 
where P: AsRef<Path>, 
```

[Implementation.](https://github.com/rust-lang/rust/compare/rust-lang:27a046e...ExpHP:path-prefix)

(**Note:** Like many similar attempts to "generalize" function signatures, this is a potentially breaking change due to potential regressions in type inference)

### PR DESCRIPTION
**Pull request by ExpHP**
I did this a while back but didn't submit a PR. Might as well see what happens.

Fixes #48390.

**Note: This has the potential to cause regressions in type inference.**  However, in order for code to break, it would need to be relying on the signature to determine that a type is `&_`, while still being able to figure out what the `_` is.  I'm having a hard time imagining such a scenario in real code.

### CODE REVIEW

**Comment by ExpHP** [comment]
Note: The [`expect` in the MultiDecorator branch](https://github.com/ExpHP/rust/blob/dc96467e220ffd68a68447a822659e245dcc9a4f/src/libsyntax/ext/expand.rs#L495) is there because, try as I might, I could not come up a test case where `ok()` failed there.  (if it ever fails in the future, it will be easier to debug now that it does not drop the `DiagnosticBuilder`)

It is only my assumption that rustc is bailing out at an earlier stage after it fails to parse the list of traits in `#[derive(...)]`.

**Comment by ExpHP** [comment]
oops, wrong PR

File: src/libstd/path.rs [hunk 1]
```diff
@@ -1876,16 +1875,19 @@ impl Path {
...
     /// assert_eq!(path.strip_prefix("test").is_ok(), false);
     /// assert_eq!(path.strip_prefix("/haha").is_ok(), false);
+    ///
+    /// let prefix = PathBuf::from("/test/");
+    /// assert_eq!(path.strip_prefix(prefix), Ok(Path::new("haha/foo.txt")));
     /// ```
     #[stable(since = "1.7.0", feature = "path_strip_prefix")]
-    pub fn strip_prefix<'a, P: ?Sized>(&'a self, base: &'a P)
-                                       -> Result<&'a Path, StripPrefixError>
+    pub fn strip_prefix<'a, P>(&'a self, base: P)
```
Comment (dtolnay) [inline src/libstd/path.rs]: I believe the `'a` lifetime can now be elided.

File: src/libstd/path.rs [hunk 2]
```diff
@@ -1876,16 +1875,19 @@ impl Path {
...
+                               -> Result<&'a Path, StripPrefixError>
         where P: AsRef<Path>
     {
         self._strip_prefix(base.as_ref())
     }
 
-    fn _strip_prefix<'a>(&'a self, base: &'a Path)
+    fn _strip_prefix<'a>(&'a self, base: &Path)
```
Comment (dtolnay) [inline src/libstd/path.rs]: Same here, I believe `'a` can be elided.

**Comment by dtolnay** [comment]
I would be on board with attempting to land this. The original signature is indeed strange, and the way `strip_prefix` is typically used I would expect nothing to break.

```diff
  impl Path {
-     pub fn strip_prefix<'a, P: ?Sized>(&'a self, base: &'a P) -> Result<&'a Path, StripPrefixError>
+     pub fn strip_prefix<P>(&self, base: P) -> Result<&Path, StripPrefixError>
          where P: AsRef<Path>;
  }
```

@rfcbot fcp merge

**Comment by dtolnay** [comment]
I would be on board with attempting to land this. The original signature is indeed strange, and the way `strip_prefix` is typically used I would expect nothing to break.

```diff
  impl Path {
-     pub fn strip_prefix<'a, P: ?Sized>(&'a self, base: &'a P) -> Result<&'a Path, StripPrefixError>
+     pub fn strip_prefix<P>(&self, base: P) -> Result<&Path, StripPrefixError>
          where P: AsRef<Path>;
  }
```

@rfcbot fcp merge

**Comment by SimonSapin** [comment]
> I'm having a hard time imagining such a scenario in real code.

This sounds like “let’s do a crater run”. What’s the process for that?

**Comment by SimonSapin** [comment]
@rfcbot concern crater

File: src/libstd/path.rs (hunk 1 above)
Comment (ExpHP) [reply src/libstd/path.rs]: Oops, thanks for catching this!

**Comment by pietroalbini** [comment]
Added this to the crater queue.

**Comment by Mark-Simulacrum** [comment]
Crater run started, ETA is ~3 days.

**Comment by Mark-Simulacrum** [comment]
Crater results all look like spurious failures. http://cargobomb-reports.s3.amazonaws.com/pr-48989/index.html

**Pull request merged**